ecosys gha (#2454)
psi4/psi4: doc/sphinxman/source/add_tests.rst

.. _`faq:add_tests`:

==========
Test Suite
==========

A test suite plays a vital role in open-source software use and development.

* For a |PSIfour| user, tests provide models of inputs that should work
  "as-is" and a searchable collection of syntax and capabilities.
  The test suite also allows high-quality development snapshots of the
  codebase to be built automatically for users.

> * For a CMS user, a test suite alongside installed |PSIfour| can be used to show that this piece is working in a complex software environment.

Reviewer: ```suggestion
* For a user who has |PSIfour| as part of a complex software environment, a test suite alongside installed |PSIfour| can be used to show that the ||PSIfour| piece is working.
```

The abbreviation "CMS" isn't standard outside of CMS, so let's use it only when necessary. If I hadn't needed to review the QCEngine paper, I'd have no clue what CMS was.

> * For a CMS user, a test suite alongside installed |PSIfour| can be used to show that this piece is working in a complex software environment.

Reviewer: ```suggestion
* For a user who has |PSIfour| as part of a complex software environment, a test suite alongside installed |PSIfour| can be used to show that the |PSIfour| piece is working.
```

The abbreviation "CMS" isn't standard outside of CMS, so let's use it only when necessary. If I hadn't needed to review the QCEngine paper, I'd have no clue what CMS was.

CTest and pytest, PSIthon and PsiAPI
====================================

In designing a test, sometimes you want it to be a model input for the user in a single file or you don't want a lot of ``psi4.`` or Python syntax cluttering the input.
In this case, follow :ref:`faq:add_psithon_tests` to prepare as PSIthon (``psi4 input.dat``) for, roughly speaking, running through :program:`ctest`.
The PSIthon/CTest test suite occupies the whole of :source:`tests` *except* :source:`tests/pytests`.

At other times you want the test to check several variations of a template job or you want to test error handling or you want to focus on PsiAPI rather than PSIthon or you want to control the compute conditions with environment variables.
In this case, follow :ref:`faq:add_psiapi_tests` to prepare as PsiAPI (``import psi4``) for, roughly speaking, running through :program:`pytest`.
The PsiAPI/pytest test suite occupies :source:`tests/pytests`.

The above description sounds as if there are two disjoint test suites, and you have to run both ``ctest`` and ``pytest`` to fully test |PSIfour|.
This has indeed been the case until March 2022.

> The difficulty has been that (1) two test suites is unexpected so some developers don't know to run both; and (2) there are important tests in the PSIthon suite that can't be run on a |PSIfour| installation since CTest only works in a build directory.

Reviewer: Does the Sphinx support footnotes? If so, this line should be one. It's important for experienced developers who want to know why things changed, but for new developers, it interrupts the line of thought.

Author: Yes, Sphinx can do footnotes, but I'd rather keep this bit all together. The appearance of two suites isn't going away directory-structure-wise, so may as well face them.

(New developers should be happy I didn't emphasize that in the source, the ctest dir contains the pytest dir, while in the installation, the pytest dir contains the ctests dirs. :-)

Reviewer: I completely agree with discussing "it looks like there are two test suites" here. But the specific thing that I think should be footnoted is the impetus for making ctests accessible through pytest.

Reviewer: This is the one item I have left on this PR. I can agree to disagree and approve regardless, but I stand by my previous comment. Whether this particular line is footnoted is unrelated to facing the appearance of two test suites.

Author: Yeah, I played with that paragraph some. I agree it can read cleaner w/o the 2-point sentence, but I didn't find a better place for the sentence. And footnote looks odd when whole docs section is 3 vertical inches. I'm inclined to leave it for now.

> The difficulty has been that (1) two test suites is unexpected so some developers don't know to run both; and (2) there are important tests in the PSIthon suite that can't be run on a |PSIfour| installation since CTest only works in a build directory.

Reviewer: Does the Sphinx support footnotes? If so, this line should be one. It's important for experienced developers who want to know why things changed, but for new developers, it interrupts the line of thought.

> Now, by adding an extra file to the test directory, PSIthon tests can also be run through :program:`pytest`.

Reviewer: Which file?

> Now, by adding an extra file to the test directory, PSIthon tests can also be run through :program:`pytest`.

Reviewer: Please link to the new section you added.

> This hasn't rolled out to all ~500 PSIthon tests (help wanted), but eventually |PSIfour| can be tested with a single command from a build or from an installation.

Reviewer: File an issue.

> This hasn't rolled out to all ~500 PSIthon tests (help wanted), but eventually |PSIfour| can be tested with a single command from a build or from an installation.

Author: done in #2495

Therefore, in designing a test, choose its mode based on whether PSIthon or PsiAPI suits it better and whether it's a simple model for users (probably PSIthon) or for expert users (probably PsiAPI).
Both will continue to work in future, and neither have limitations.

Test Contents
=============

> * If you're writing a new class or closely examining an old one and you develop unit tests for probing its behavior, by all means add them to the test suite.
>   But on the whole, |PSIfour| tests should be integration tests focusing on non-regression of user input to answers.

Reviewer: ```suggestion
* Most tests will be |PSIfour| tests integration tests focusing on non-regression of user input to answers, and we insist on having these.
  But if you find unit tests helpful, by all means add them to the test suite.
```

Put the important part first, remove the assumption about unit testing being for a class, and clarify what we demand for tests vs. what we predict for tests.

> * If you're writing a new class or closely examining an old one and you develop unit tests for probing its behavior, by all means add them to the test suite.
>   But on the whole, |PSIfour| tests should be integration tests focusing on non-regression of user input to answers.

Reviewer: ```suggestion
* If you're writing a new class or closely examining an old one and you develop unit tests for probing its behavior, by all means add them to the test suite.
  But on the whole, |PSIfour| tests should be integration tests focusing on non-regression of user input to answers.
```
```suggestion
* On the whole, |PSIfour| tests should be integration tests focusing on non-regression of user input to answers.
   But if you find unit tests helpful, by all means add them to the test suite.
```

> * General format is to store reference results (from literature or another implementation or a carefully run |PSIfour| calculation),
>   run quantum chemistry, then apply one or more of the :ref:`faq:comparison_functions` so that the test will fail if the answer is unexpected.
>   The functions are the same in CTest and pytest, but in the former it's standalone ``compare_matrices(refmat, mat, ...)`` while in the latter it's asserted ``assert compare_matrices(refmat, mat, ...)``.

Reviewer: ```suggestion
* Most tests should store reference results (from literature or another implementation or a carefully run |PSIfour| calculation),
  run quantum chemistry, then apply one or more of the :ref:`faq:comparison_functions` so that the test will fail if the answer is unexpected.
  The functions are the same in CTest and pytest, but in the former they are, for example, ``compare_matrices(refmat, mat, ...)`` while in the latter it's asserted, like ``assert compare_matrices(refmat, mat, ...)``.
```

> * General format is to store reference results (from literature or another implementation or a carefully run |PSIfour| calculation),
>   run quantum chemistry, then apply one or more of the :ref:`faq:comparison_functions` so that the test will fail if the answer is unexpected.
>   The functions are the same in CTest and pytest, but in the former it's standalone ``compare_matrices(refmat, mat, ...)`` while in the latter it's asserted ``assert compare_matrices(refmat, mat, ...)``.

Reviewer: ```suggestion
* Most tests should store reference results (from literature or another implementation or a carefully run |PSIfour| calculation),
  run quantum chemistry, then apply one or more of the :ref:`faq:comparison_functions` so that the test will fail if the answer is unexpected.
  The functions are the same in CTest and pytest, but in the former they are, for example, ``compare_matrices(refmat, mat, ...)`` while in the latter it's asserted, like ``assert compare_matrices(refmat, mat, ...)``.
```

  or the default convergence for the calculation type (energy, gradient, *etc.*).

* Keep tests as short as possible without sacrificing coverage and variety. Under 30 seconds is a good aim.

.. _`faq:add_psithon_tests`:

Adding PSIthon Test Cases
=========================

To create a new test case, first make a folder in :source:`tests` or, for an addon, a subfolder under the addon folder.
Use hyphens, not spaces or underscores, in the directory name.
Add the directory name to the list of tests in :source:`tests/CMakeLists.txt` or, for an addon, ``tests/<addon>/CMakeLists.txt``.
The test directory will need at least two files, ``CMakeLists.txt`` and ``input.dat``.

``CMakeLists.txt``
------------------

This file adds the test case to the suite. It should have at least the following two uncommented lines::

    include(TestingMacros)

    # if extra files needed
    # file(COPY grid.dat DESTINATION ${CMAKE_CURRENT_BINARY_DIR})

    add_regression_test(directory_name "psi;semicolon_separated-list-of-applicable-test-labels")

>     # if minutest long

Reviewer: Minutest?

>     # if minutest long

Reviewer: Minutest?

    # set_tests_properties(isapt1 PROPERTIES COST 300)

The labels specify which groups of tests include the test case for ``ctest -L label`` purposes. The ``psi`` label should always be added, but the other labels are test-specific. The method tested should always be included, and this is often sufficient. If adding a test for an already existing module, the labels for other tests of the module will suggest other labels to add.

> Labels have been added as developers needed, so they are not systematic or thorough. Help wanted.

Reviewer: ```suggestion
Labels have been added as developers needed, so they are not systematic or thorough. If you see labels to add or rename, please do.
```

> Labels have been added as developers needed, so they are not systematic or thorough. Help wanted.

Reviewer: ```suggestion
Labels have been added as developers needed, so they are not systematic or thorough. If you see labels to add or rename, please do.
```

A test requiring over 15 minutes should be labeled ``longtests``. A short test under 30 seconds used for general bug checking should be labeled ``quicktests``. A test that confirms |PSIfour| is operational should be labeled ``smoketests``.

> If a test needs extra input files like ``grid.dat`` or extra reference files for checking against like ``fchk``, specify these in the ``CMakeLists.txt`` as shown above. Such tests must be run through ``ctest`` and don't usually work when run "by hand" from the objdir via ``stage/bin/psi4 ../tests/directory_name/input.dat``.

Reviewer: ```suggestion
If a test needs extra input files like ``grid.dat`` or extra reference files for checking against like ``fchk``, specify these in the ``CMakeLists.txt`` as shown above. Such tests must be run through ``ctest`` and don't usually work when run "by hand" from the objdir via ``stage/bin/psi4 ../tests/directory_name/input.dat``.
```
```suggestion
If a test needs extra input files like ``grid.dat`` or extra reference files for checking against, like ``fchk``, specify these in the ``CMakeLists.txt`` as shown above. Such tests must be run through ``ctest`` and don't usually work when run "by hand" from the objdir via ``stage/bin/psi4 ../tests/directory_name/input.dat``.
```

    compare_values(refenergy, thisenergy, 9, "Reference energy")  #TEST
    compare_values(refenergy, variable('scf total energy'), 9, "Reference energy")  #TEST

Of those small modifications, first, note the special comment at the top (starting with the ``#!`` comment marker). This should be descriptive since it is inlined into the manual (unless ``!nosample`` in the comment) as a sample input.

Reference values are often assigned to variables for later use.
The compare_values function (along with several relatives in :source:`psi4/driver/p4util/testing.py` for comparing strings, matrices, etc.) checks that the computed values match these reference values to suitable precision. This function prints an error message and signals that the test failed to the make system, if the values don't match. Any lines of the input associated with the validation process should be flagged with ``#TEST`` at the end of each line, so that they can be removed when copying from the tests to the samples directory.

``output.ref``
--------------

When your test case is in final form, run it locally, rename the output to ``output.ref``, and check it into the repository alongside ``input.dat``.

> While this isn't used for any testing machinery (except for the nearly decommissioned :source:`tests/psitest.pl` for CC tests), it can be handy for users or developers to consult.

Reviewer: I am hoping for full decommission by 1.6. 🙂

> While this isn't used for any testing machinery (except for the nearly decommissioned :source:`tests/psitest.pl` for CC tests), it can be handy for users or developers to consult.

Reviewer: I am hoping for full decommission by 1.6. 🙂

* ``smoke`` used instead of CTest ``smoketests``
* ``quick`` used instead of CTest ``quicktests``
* ``long`` used instead of CTest ``longtests``
* ``addon`` and ``<name-of-addon>`` added automatically when ``@uusing("<name-of-addon>")`` decorates the test or ``marks=using("<name-of-addon>")`` marks the test

CTest "labels" are called "marks" in pytest.
Any new marks should be added to :source:`pytest.ini`.

Running for Debugging
---------------------

* PSIthon tests that don't need extra files to run are easily run from ``<objdir>`` via ``stage/bin/psi4 ../tests/<test-name>/input.dat, with the output appearing in ``../tests/<test-name>/input.out``.
* All PSIthon tests are runable through CTest, and output files appear in ``<objdir>/tests/<test-name>/output.dat`` and stdout results appear in ``<objdir>/Testing/Temporary/LastTest.log*``.

.. _`faq:add_psiapi_tests`:

Adding PsiAPI Test Cases
========================

To create a new test case, either create a new file or add to an existing file under :source:`tests/pytests`.

* Test must be in the :source:`tests/pytests/` directory.
* Test file name must start with ``test_``. This is how pytest knows to collect it.

> * Test file may contain many tests, each of which is an ordinary Python function with name starting ``test_``. This is how pytest knows to collect it.

Reviewer: ```suggestion
* A test file may contain many tests, each of which is an ordinary Python function with name starting ``test_``. This is how pytest knows to collect it.
```

* No registration required to bring a test to pytest's attention.

> * No registration required to bring a test to CMake's attention. If a test needs additional files, register them in :source:`psi4/CMakeLists.txt`.

Reviewer: Huh? This is the pytest section, so why are we bringing in cmake?

Author: Ah yes, pytests escape _CTest_'s attention but because they should be installed for users (not just run from source for developers), _CMake_ is still involved at the `install(FILES` or `install(DIRECTORY` level.

A few notes on test contents:

* Import testing functions from ``utils`` and use Python assert: ``assert compare_values(expected, computed, ...)``.

> * Don't worry about cleaning up files or resetting options. There's a function in :source:`tests/pytests/conftest.py` that does so between every test.

Reviewer: ```suggestion
* Don't worry about cleaning up files or resetting options. There's a function in :source:`tests/pytests/conftest.py` that does so between every test.
```
```suggestion
* Don't worry about cleaning up files or resetting options. A function in :source:`tests/pytests/conftest.py` does this automatically.
```

> * Especially if using data or functions from outside a test, run a variety of tests at different parallelisms. If tests fail that pass when run alone, you've got a function of the same name changing state or some similar correctable phenomenon.

Reviewer: What is "parallelism"? Based on the example you give, I don't think you mean "number of threads."

Author: I think it's multiprocessing vs. multithreading. (We could also use the latter, but that's off topic.) Edited to 

```
tests at different ``pytest -n <N>`` levels to mix up test ordering. If tests fail ...
```

> * Especially if using data or functions from outside a test, run a variety of tests at different parallelisms. If tests fail that pass when run alone, you've got a function of the same name changing state or some similar correctable phenomenon.

Reviewer: What is "parallelism"? Based on the example you give, I don't think you mean "number of threads."
